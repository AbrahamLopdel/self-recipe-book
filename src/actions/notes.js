import Swal from "sweetalert2";
import { db } from "../firebase/firebaseConfig";
import { fileUpload } from "../helpers/fileUpload";
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";

export const startNewNote = () => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;

    const newNote = {
      title: "",
      body: "",
      steps: "",
      food: "",
      time: 0,
      vegetarian: false,
      date: new Date().getTime(),
    };
    try {
      const document = await db.collection(`${uid}/journal/notes`).add(newNote);

      dispatch(activeNote(document.id, newNote));
      dispatch(addNewNote(document.id, newNote));
    } catch (error) {
      console.log(error);
    }
  };
};

export const activeNote = (id, note) => ({
  type: types.noteActive,
  payload: {
    id,
    ...note,
  },
});

export const noActiveNote = () => ({
  type: types.notesNoActive,
  payload: null,
});

export const addNewNote = (id, note) => ({
  type: types.notesAddNew,
  payload: {
    id,
    ...note,
  },
});

export const startLoadingNotes = (uid) => {
  return async (dispatch) => {
    const notes = await loadNotes(uid);
    dispatch(setNotes(notes));
  };
};

export const setNotes = (notes) => ({
  type: types.notesLoad,
  payload: notes,
});

export const cancelActiveNote = (note) => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;

    const newNote = {
      title: "",
      body: "",
      steps: "",
      food: "",
    };

    if(note.title === newNote.title && note.body === newNote.body && note.steps === newNote.steps && note.food === newNote.food ) {
      await db.doc(`/${uid}/journal/notes/${note.id}`).delete();
      dispatch(deleteNote(note.id))
    }

    dispatch(noActiveNote());
  }
}

export const startSaveNote = (note) => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;

    if (!note.url) {
      delete note.url;
    }

    const noteToFirestore = { ...note };
    delete noteToFirestore.id;

    const newNote = {
      title: "",
      body: "",
      steps: "",
      food: "",
    };

    if(note.title === newNote.title && note.body === newNote.body && note.steps === newNote.steps && note.food === newNote.food ) {
      await db.doc(`/${uid}/journal/notes/${note.id}`).delete();
      dispatch(deleteNote(note.id))
    }
    else {
      await db.doc(`/${uid}/journal/notes/${note.id}`).update(noteToFirestore);
    }

    dispatch(refreshNote(note.id, noteToFirestore));
    Swal.fire("Saved", note.title, "success");
  };
};

export const refreshNote = (id, note) => ({
  type: types.noteUpdated,
  payload: {
    id,
    note: {
      id,
      ...note,
    },
  },
});

export const startUploading = (file) => {
  return async (dispatch, getState) => {
    const { active: activeNote } = getState().notes;

    Swal.fire({
      title: "Uploading...",
      text: "Please wait...",
      allowOutsideClick: false,
      showConfirmButton: false,
      willOpen: () => {
        Swal.showLoading();
      },
    });

    const fileUrl = await fileUpload(file);
    activeNote.url = fileUrl;

    dispatch(startSaveNote(activeNote));
    Swal.close();
  };
};

export const startDeleting = (id) => {
  return async (dispatch, getState) => {
    const note = getState().notes.active
    const result = await Swal.fire({
      title: `¿Estás seguro de que deseas eliminar la receta "${
        note.title ? note.title : "Sin nombre"
      }"?`,
      showCancelButton: true,
      confirmButtonText: `Eliminar`,
    });

    if (result.isConfirmed) {
      const uid = getState().auth.uid;

      await db.doc(`/${uid}/journal/notes/${id}`).delete();

      dispatch(deleteNote(id));
    }
  };
};

export const deleteNote = (id) => ({
  type: types.notesDelete,
  payload: id,
});

export const notesLogout = () => ({
  type: types.notesLogoutCleaning,
});

export const notesFilteredByTitle = (notesName) => ({
  type: types.notesFilteredByTitle,
  payload: notesName,
});

export const notesFilteredByBody = (notesDesc) => ({
  type: types.notesFilteredByBody,
  payload: notesDesc,
});

export const notesFilteredByFood = (notesFood) => ({
  type: types.notesFilteredByFood,
  payload: notesFood,
});

export const notesFilteredByTime = (notesTime) => ({
  type: types.notesFilteredByTime,
  payload: notesTime,
});

export const notesFilteredBy = (notes, filter) => ({
  type: types.notesFilteredBy,
  payload: notes,
  filter,
});
