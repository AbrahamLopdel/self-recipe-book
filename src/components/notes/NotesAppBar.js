import moment from "moment";
import React from "react";
import { useDispatch, useSelector } from "react-redux";
import {
  cancelActiveNote,
  startSaveNote,
  startUploading,
} from "../../actions/notes";

export const NotesAppBar = ({ history }) => {
  const dispatch = useDispatch();
  const { active } = useSelector((state) => state.notes);
  
  const noteDate = moment(active.date);

  const handleSave = () => {
    dispatch(startSaveNote(active));
  };

  const handlePictureUpload = () => {
    document.querySelector("#fileSelector").click();
  };

  const handleFileChange = (e) => {
    const file = e.target.files[0];
    if (file) {
      dispatch(startUploading(file));
    }
  };

  const handleCancel = () => {
    dispatch(cancelActiveNote(active));
  };

  return (
    <div className="notes__appbar">
      <i
        className="far fa-window-close notes__cancel"
        onClick={handleCancel}
      ></i>
    <span>{noteDate.format("MMMM Do YYYY")}</span>

      <input
        id="fileSelector"
        type="file"
        name="file"
        style={{ display: "none" }}
        onChange={handleFileChange}
      />

      <div>
        <button className="btn" onClick={handlePictureUpload}>
          Picture
        </button>
        <button className="btn" onClick={handleSave}>
          Save
        </button>
      </div>
    </div>
  );
};
