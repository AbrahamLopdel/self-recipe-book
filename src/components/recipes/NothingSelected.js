import React from "react";

export const NothingSelected = () => {
  return (
    <div className="nothing__main-content animate__animated animate__fadeIn animate__faster">
      <p>
        Selecciona o crea
        <br />
        alguna receta
      </p>
      <i className="far fa-star fa-4x mt-5"></i>
    </div>
  );
};
