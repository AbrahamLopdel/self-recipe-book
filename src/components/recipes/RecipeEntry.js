import React from "react";
import moment from "moment";
import { useDispatch } from "react-redux";
import { activeNote } from "../../actions/notes";

export const RecipeEntry = ({
  id,
  date,
  food,
  title,
  time,
  body,
  steps,
  vegetarian,
  url,
}) => {
  const dispatch = useDispatch();

  const noteDate = moment(date);

  const note = {
    date,
    food,
    title,
    time,
    body,
    steps,
    vegetarian,
    url,
  };

  const handleEntryClick = () => {
    dispatch(activeNote(id, note));
  };

  return (
    <div className="randomClass" onClick={handleEntryClick}>
      <div className="recipes__entry pointer animate__animated animate__bounceInLeft animate__faster">
        {url && (
          <div
            className="recipes__entry-picture"
            style={{
              backgroundSize: "cover",
              backgroundImage: `url(${url})`,
            }}
          ></div>
        )}

        <div className="recipes__entry-body">
          <p className="recipes__entry-title">{title}</p>
        </div>

        <div className="recipes__entry-date-box">
          <span>{noteDate.format("dddd")}</span>
          <h4>{noteDate.format("DD")}</h4>
        </div>
      </div>
    </div>
  );
};
