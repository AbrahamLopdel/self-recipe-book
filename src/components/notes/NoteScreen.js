import React, { useEffect, useRef, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { activeNote, startDeleting } from "../../actions/notes";
import { useForm } from "../../hooks/useForm";
import { NotesAppBar } from "./NotesAppBar";

export const NoteScreen = ({ history }) => {
  const dispatch = useDispatch();
  const { active: entry } = useSelector((state) => state.notes);
  const [foodFocus, setFoodFocus] = useState(false);
  const [foodAr, setFoodAr] = useState([]);
  const [vegetarian, setVegetarian] = useState(entry.vegetarian);
  const [values, handleInputChange, reset] = useForm(entry);
  const { body, title, steps, time, food } = values;

  const activeId = useRef(entry.id);
  const [nextStep, setNextStep] = useState(1);
  

  useEffect(() => {
    if (vegetarian) {
      document.querySelector("#vg").checked = true;
    }
  }, [vegetarian]);

  useEffect(() => {
    if (entry.id !== activeId.current) {
      reset(entry);
      activeId.current = entry.id;
    }
  }, [entry, reset]);

  useEffect(() => {
    if (food) setFoodAr(food.split(","));
  }, [food]);

  useEffect(() => {
    dispatch(activeNote(values.id, { ...values, vegetarian }));
  }, [values, vegetarian, dispatch]);

  function handleSteps(e) {
    if(e.key === 'Enter') {
      setNextStep(nextStep + 1);
      document.querySelector("#steps").value = nextStep + ". ";
    }
  }

  const handleVegetarian = () => {
    setVegetarian(!vegetarian);
  };

  const handleFoodFocus = () => {
    setFoodFocus(!foodFocus);
  };

  const handleDelete = () => {
    dispatch(startDeleting(entry.id));
  };

  return (
    <div className="notes__main-content animate__animated animate__fadeIn animate__faster">
      <NotesAppBar history={history} />

      <div className="notes__content">
        <input
          type="text"
          placeholder="Título de la receta"
          className="notes__title-input"
          autoComplete="off"
          name="title"
          value={title}
          onChange={handleInputChange}
        />
        <textarea
          placeholder="Descripción de la receta"
          className="notes__textarea notes__textarea-ingredientes"
          name="body"
          value={body}
          onChange={handleInputChange}
        ></textarea>
        <textarea
          id="steps"
          placeholder="Pasos a seguir"
          className="notes__textarea notes__textarea-ingredientes"
          name="steps"
          value={steps}
          onChange={handleInputChange}
          onKeyDown={handleSteps}
        ></textarea>
        <label className="notes__food">
          <button
            className={`btn btn-primary btn-focus ${
              foodFocus && "btn-success"
            }`}
            onClick={handleFoodFocus}
          >
            Ingredientes:
          </button>
          (Poner ingredientes seguidos de comas)
          {foodFocus ? (
            <textarea
              placeholder="Ingrediente"
              id="ingredientes_textarea"
              className="notes__time-input"
              name="food"
              value={food}
              onChange={handleInputChange}
            ></textarea>
          ) : (
            <ul>
              {foodAr.map((food, i) => {
                return <li key={i}>{food}</li>;
              })}
            </ul>
          )}
        </label>
        <label>
          Duración:
          <input
            type="number"
            placeholder="0"
            className="notes__time-input"
            autoComplete="off"
            name="time"
            value={time}
            onChange={handleInputChange}
          />
        </label>
        <div className="form-check form-switch ">
          <label className="form-check-label">
            Vegetariana
            <input
              className="form-check-input"
              type="checkbox"
              id="vg"
              name="vegetarian"
              value={vegetarian}
              onClick={handleVegetarian}
            />
          </label>
        </div>
        {entry.url && (
          <div className="notes__image">
            <img src={entry.url} alt="imagen" />
          </div>
        )}
      </div>

      <button className="btn btn-danger btn-delete" onClick={handleDelete}>
        Delete
      </button>
    </div>
  );
};
