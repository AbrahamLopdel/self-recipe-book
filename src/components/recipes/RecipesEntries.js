import React from "react";
import { useSelector } from "react-redux";
import { RecipeEntry } from "./RecipeEntry";

export const RecipesEntries = () => {
  const { notes } = useSelector((state) => state.notes);

  return (
    <div className="recipes__entries">
      {(notes.length > 0) ? 
      notes.map((note) => (
        <RecipeEntry key={note.id} {...note} />
      )) :
      (<div className="recipes__empty">
        <p>Todavía no tienes recetas en tu recetario</p>
        <p>¡Crea una cuanto antes!</p>
      </div>)}
    </div>
  );
};
