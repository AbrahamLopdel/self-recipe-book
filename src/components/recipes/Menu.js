import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { startLogout } from "../../actions/auth";
import { startNewNote } from "../../actions/notes";
import { RecipesEntries } from "./RecipesEntries";

export const Menu = ({ history }) => {
  const dispatch = useDispatch();

  const { name } = useSelector((state) => state.auth);

  const handleLogout = () => {
    dispatch(startLogout());
  };

  const handleAddNew = () => {
    dispatch(startNewNote());
  };

  const handleSearchRecipe = () => {
    history.push("/search-recipe");
  };

  return (
    <section className="recipes__sidebar">
      <div className="recipes__sidebar-navbar">
        <h3 className="mt-5 ml-5">
          <i className="far fa-moon"></i>
          <span> {name}</span>
        </h3>

        <button className="btn" onClick={handleLogout}>
          Logout
        </button>
      </div>
      <div className="recipes__submenu">
        <div className="recipes__search-entry" onClick={handleSearchRecipe}>
          <i className="fas fa-search fa-1x"></i>
          <p className="mt-1">Buscar recetas</p>
        </div>
        <div className="recipes__new-entry" onClick={handleAddNew}>
          <i className="far fa-calendar-plus fa-1x"></i>
          <p className="mt-1">Nueva receta</p>
        </div>
      </div>

      <RecipesEntries />
    </section>
  );
};
