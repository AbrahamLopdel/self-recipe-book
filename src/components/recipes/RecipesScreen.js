import React from "react";
import { Menu } from "./Menu";
// import { NothingSelected } from "./NothingSelected";
import { NoteScreen } from "../notes/NoteScreen";
import { useSelector } from "react-redux";

export const RecipesScreen = ({ history }) => {
  const { active } = useSelector((state) => state.notes);

  return (
    <div className="recipes__main-content animate__animated animate__fadeIn">
      <main>
        {active ? <NoteScreen history={history} /> : <Menu history={history} />}
      </main>
    </div>
  );
};
