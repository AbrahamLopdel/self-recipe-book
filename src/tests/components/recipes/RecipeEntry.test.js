import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store"; //ES6 modules
import thunk from "redux-thunk";
import { activeNote } from "../../../actions/notes";
import { RecipeEntry } from "../../../components/recipes/RecipeEntry";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initialState = {};

let store = mockStore(initialState);

store.dispatch = jest.fn();

describe("Pruebas en RecipeEntry", () => {
  const note = {
    id: 10,
    date: 0,
    title: "Ayuda pofavo",
    body: "equisde",
    url: "https://flaticon.com/mouth.png",
  };

  const wrapper = mount(
    <Provider store={store}>
      <RecipeEntry {...note} />
    </Provider>
  );

  test("Debe de mostrarse correctamente", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("Debe de disparar el activeNote", () => {
    wrapper.find(".randomClass").prop("onClick")();

    expect(store.dispatch).toHaveBeenCalledWith(
      activeNote(note.id, { ...note })
    );
  });
});
