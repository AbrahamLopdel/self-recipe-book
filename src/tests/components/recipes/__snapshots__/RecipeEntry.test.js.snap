// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en RecipeEntry Debe de mostrarse correctamente 1`] = `
<div
  className="randomClass"
  onClick={[Function]}
>
  <div
    className="recipes__entry pointer animate__animated animate__bounceInLeft animate__faster"
  >
    <div
      className="recipes__entry-picture"
      style={
        Object {
          "backgroundImage": "url(https://flaticon.com/mouth.png)",
          "backgroundSize": "cover",
        }
      }
    />
    <div
      className="recipes__entry-body"
    >
      <p
        className="recipes__entry-title"
      >
        Ayuda pofavo
      </p>
      <p
        className="recipes__entry-content"
      >
        equisde
      </p>
    </div>
    <div
      className="recipes__entry-date-box"
    >
      <span>
        Thursday
      </span>
      <h4>
        01
      </h4>
    </div>
  </div>
</div>
`;
